### PR TITLE
[Feature][HunyuanImage3.0] Add cfgP to HunyuanImage3.0

### PR DESCRIPTION
**Pull request by nussejzz**
## Purpose
 
Add CFG (Classifier-Free Guidance) parallel support for HunyuanImage3.0.
 
HunyuanImage3.0 previously ran CFG sequentially by doubling the batch (`torch.cat([latents] * 2)`) and processing conditioned + unconditioned branches in a single forward pass. This PR enables distributing the two branches across separate GPUs via the existing CFG parallel infrastructure (`get_cfg_group`, `get_classifier_free_guidance_rank`).

### `hunyuan_image_3_transformer.py`
 
- **`_split_model_kwargs_for_cfg_parallel`**: Splits batch-doubled model kwargs (produced by the tokenizer with `cfg_factor=2`) so that rank 0 processes the conditioned branch and rank 1 processes the unconditioned branch. Handles tensors, `custom_pos_emb` (cos/sin tuple), `cond_vae_images`, `cond_vit_images`, `vit_kwargs`, etc.
- **Modified `HunyuanImage3Text2ImagePipeline.__call__`**:
  - Detects CFG parallel via `get_classifier_free_guidance_world_size() == 2`
  - Broadcasts initial latents to ensure noise consistency across ranks
  - Splits `input_ids`, `attention_mask`, and all batch-dim kwargs before the denoising loop
  - Each rank runs `forward_call` with `batch_size=1` independently, maintaining its own KV cache
  - Uses `all_gather` → rank 0 combines + scheduler step → `broadcast` latents (standard 2-branch pattern, consistent with `CFGParallelMixin`)
  - Casts latents to `bfloat16` before broadcast to prevent dtype mismatch (scheduler step upcasts to `float32`, which corrupts the `bfloat16` buffer on rank 1 during broadcast)
 
### `pipeline_hunyuan_image_3.py`
 
- **Device mapping fix**: Replaced `tp_rank`-based device assignment with `get_local_device()` for non-TP modules (`vae`, `time_embed`, `patch_embed`, `final_layer`, etc.). In CFG parallel mode, `tp_rank` resets to 0 within each CFG group, causing modules on CFG rank 1 to be placed on the wrong GPU (e.g., `cuda:0` instead of `cuda:4`). `get_local_device()` returns the actual CUDA device via `LOCAL_RANK`, fixing this for both CFG parallel and non-CFG parallel setups.
 
### `hunyuan_image3_moe_dit.yaml`
 
- Updated stage config to enable CFG parallel (`cfg_parallel_size: 2`) with `tensor_parallel_size: 4` on 8 GPUs.
 
## Why Not Reuse `CFGParallelMixin`?
 
HunyuanImage3.0 uses a native multimodal architecture with autoregressive KV cache management (`_save_image_kv_caches`, `_update_model_kwargs_for_generation`), which doesn't fit the `predict_noise(**kwargs) → tensor` abstraction that `CFGParallelMixin` assumes. The communication pattern (`all_gather` + `broadcast`) is identical to `CFGParallelMixin`, but the complex forward pass with KV cache management is implemented inline.
 
## Backward Compatibility
 
When `cfg_parallel_size == 1`, the original sequential CFG path is preserved unchanged. The `get_local_device()` fix is also safe for non-CFG-parallel setups since `LOCAL_RANK == tp_rank` when there is no CFG parallel.
 
## Test Plan

`python3 examples/offline_inference/text_to_image/text_to_image.py \
  --model /data/HunyuanImage-3.0 \
  --stage-configs-path vllm_omni/model_executor/stage_configs/hunyuan_image3_moe_dit.yaml \
  --prompt "A brown and white dog is running on the grass" `
 
- ✅ TP4 + CFG2 (8x A100): Verify image generation produces valid output (no NaN)
- ✅ TP8 (8x A100): Verify baseline still works correctly
- ✅ TP4 (4x A100): Verify non-CFG-parallel path is unaffected
- ✅ Performance benchmark: Compare TP4+CFG2 vs TP8 vs TP4
 
## Test Results
 
Benchmarked on 8x A100, HunyuanImage3.0, 1024x1024, 50 inference steps:

### TP4
image-20260324150325483
image-20260324150130470

### TP8

image-20260324144519595

image-20260324145401086

### TP4 + CFG2

image-20260324144906598

image-20260324145230487

### Summary
 
| Configuration | GPUs | Avg Step Time | Total Denoise | Speedup vs TP4 |
|---------------|------|---------------|---------------|----------------|
| TP4           | 4    | 0.535 s       | 26.8 s        | 1.00x (baseline) |
| TP8           | 8    | 0.380 s       | 19.0 s        | 1.41x          |
| TP4 + CFG2    | 8    | 0.306 s       | 15.3 s        | 1.75x          |
 
CFG parallel achieves **1.75x speedup** over TP4 and **20% speedup** over TP8 with the same 8 GPUs. CFG inter-rank communication overhead is only **0.8%** of total denoising time.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [x] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by Gaohan123** [comment]
@princepride @wtomin PTAL

**Comment by wtomin** [comment]
#1832 maintains a table for e2e test。 Would you like contribute to hunyuanimage3 e2e test? You can start with a small test case coverage that at least includes cfg parallel.

**Comment by nussejzz** [comment]
> #1832 maintains a table for e2e test。 Would you like contribute to hunyuanimage3 e2e test? You can start with a small test case coverage that at least includes cfg parallel.

Yes of course! @wtomin

**Comment by nussejzz** [comment]
PTAL!  @Bounty-hunter 😊

**Comment by Bounty-hunter** [comment]
> PTAL! @Bounty-hunter 😊

LGTM

File: vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py [hunk 1]
```diff
@@ -2416,14 +2500,25 @@ def __call__(
...
+                    # Ensure consistent dtype before broadcast (scheduler may upcast to float32)
+                    latents = latents.to(dtype=torch.bfloat16).contiguous()
+                    cfg_group.broadcast(latents, src=0)
+                else:
+                    if self.do_classifier_free_guidance:
+                        pred_cond, pred_uncond = pred.chunk(2)
+                        pred = self.cfg_operator(pred_cond, pred_uncond, self.guidance_scale, step=i)
+                    latents = self.scheduler.step(pred, t, latents, **_scheduler_step_extra_kwargs, return_dict=False)[
+                        0
+                    ]
```
Comment (princepride) [inline vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py]: Can we directly use `vllm_omni/diffusion/distributed/cfg_parallel`, if can't, can we use `all_cfg_parallel[all_gather] -> calculate` instead of `rank_0_cfg_parallel[all_gather] -> calculate -> boardcast`

Comment (nussejzz) [reply vllm_omni/diffusion/models/hunyuan_image_3/hunyuan_image_3_transformer.py]: Thanks for the review!

**Why not reuse `CFGParallelMixin`**: HunyuanImage3 maintains an autoregressive KV cache across denoising steps — after each step, `_update_model_kwargs_for_generation` threads updated KV state into the next iteration. This per-rank stateful management doesn't fit the stateless `predict_noise(**kwargs)` interface that `CFGParallelMixin` is built around.

**On the current pattern**: `all_gather → rank-0 combine + scheduler step → broadcast` is exactly the same logic as `CFGParallelMixin.scheduler_step_maybe_with_cfg`, so we're consistent with the existing convention in the repo.

**On having all ranks calculate independently (to drop the broadcast)**: This doesn't work reliably for MoE models. Unlike dense architectures, expert routing introduces non-determinism — different runs may dispatch to different experts, meaning rank-0 and rank-1 could produce slightly different outputs even given identical inputs. The broadcast ensures both ranks start the next step from the exact same latents.

**Comment by Bounty-hunter** [comment]
I test with:
```
python examples/offline_inference/text_to_image/text_to_image.py --model /data/models/tencent/HunyuanImage-3-Instruct --prompt "A brown and white dog is running on the grass" --output output_without_tp.png --num-inference-steps 50 --guidance-scale 5.0 --seed 1234
```

**H800**
| case | time  |output |
|-----|-----|-----|
| TP4 |  13.4153 s | output_tp4 |
| TP2-cfgp2 | 9.7 s | output_tp2_cfg2 |

**Comment by princepride** [comment]
@nussejzz Can you test the implementation that cfg_p without boadcast, I want know whether it will affect result.

**Comment by nussejzz** [comment]
> @nussejzz Can you test the implementation that cfg_p without boadcast, I want know whether it will affect result.

The broadcast in the denoising loop has been removed — now using all_gather → local combine on all ranks, consistent with CFGParallelMixin. The only remaining broadcast is before the loop to sync initial random latents across ranks. @princepride
